### ISSUE TITLE
Dart roll has introduced analyzer integration problems

### ISSUE THREAD

**Issue by gaaclarke** [flutter/flutter#96931]
flutter is seeing unit test failures that are asserting the output of the analyzer after a dart roll.

Here is the failed roll: https://github.com/flutter/flutter/pull/96906

Here is the range of Dart commits that probably caused it: https://dart.googlesource.com/sdk.git/+log/b3f3c7dbd422..ff0b0405c13a  

Here is a link to the failures: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8824517516265048801/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_2_4/test_stdout 

Here is an example failure: 
```
05:43 +50 -1: test/integration.shard/analyze_once_test.dart: working directory with errors [E]                                                                                                         
  Expected: contains 'warning • The parameter \'onPressed\' is required'
    Actual: 'Analyzing flutter_project...                                    \n'
              '\n'
              '   info • Avoid empty else statements • lib/main.dart:8:48 • avoid_empty_else\n'
              '   info • Avoid empty statements • lib/main.dart:8:48 • empty_statements\n'
              '   info • The declaration \'_incrementCounter\' isn\'t referenced • lib/main.dart:30:8 • unused_element\n'
              '\n'
              ''
  
  package:test_api                                     expect
  test/integration.shard/analyze_once_test.dart 314:7  assertContains
  test/integration.shard/analyze_once_test.dart 33:5   main.runCommand
  ===== asynchronous gap ===========================
  test/integration.shard/analyze_once_test.dart 135:5  main.<fn>
  

05:43 +50 -1: loading test/integration.shard/analyze_once_test.dart                                                                                                                                    
Output of flutter analyze --no-pub

Analyzing flutter_project...                                    

   info • Avoid empty else statements • lib/main.dart:8:48 • avoid_empty_else
   info • Avoid empty statements • lib/main.dart:8:48 • empty_statements
   info • The declaration '_incrementCounter' isn't referenced • lib/main.dart:30:8 • unused_element

3 issues found. (ran in 1.7s)
```

**Comment by DanTup** [comment]
I was able to reproduce this issue locally by:

- Changing my local Flutter install to have 5840dab81ea8b244fe60f4f9ac43058909c3a681 as the version hash in bin/internal/engine.version
- Running "flutter doctor" to fetch that engine
- Running the test from the `flutter_tools` folder using `../../bin/cache/dart-sdk/bin/pub run test --test-randomize-ordering-seed=20220120 -rcompact -j1 --no-color test/integration.shard/analyze_once_test.dart`

I tweaked the test to not delete the folder after running the test and opened it in VS Code, and I saw the same thing - there was no complaint about the missing onPressed argument:

<img width="673" alt="Screenshot 2022-01-20 at 19 45 51" src="https://user-images.githubusercontent.com/1078012/150411413-11c5c38f-655d-4052-b754-b456c026ddf6.png">

When I look in `.dart_tool/package_config.json`, I see the Flutter version marked as 2.10:

<img width="772" alt="Screenshot 2022-01-20 at 19 46 38" src="https://user-images.githubusercontent.com/1078012/150411446-046fd31d-6a06-4abe-b9ad-2ef7070bee2b.png">

If I modify the `pubspec.yaml` so `pub get` runs, this version number is updated to 2.12, and then the `onPressed` error appears.

So to me, it seems like this behaviour may be correct if the `required` keyword is being ignored in `flutter` because its being treated as 2.10. However this doesn't explain why it would have ever worked, so perhaps I'm missing something.

@bwilkerson @scheglov any ideas what might be happening here (or how any of the changes in https://dart.googlesource.com/sdk.git/+log/b3f3c7dbd422..ff0b0405c13a may have changed the behaviour)?

**Comment by scheglov** [comment]
It might be caused by https://github.com/dart-lang/sdk/commit/6d5543e01bd68099e068a3a88c48a282edf828e7, `non-nullable` is not an experiment anymore.

**Comment by scheglov** [comment]
So, `2.10` was `experimentalReleaseVersion` for `non-nullable`, and it was possible to enable it as an experiment. And because the experiment is over now, and the feature released in `2.12`, you have to have at least `2.12` to get it.

**Comment by DanTup** [comment]
Ah, that explains it. I've verified changing to 2.12 fixes the test for me locally. I've opened a PR at https://github.com/flutter/flutter/pull/96950. Assuming it's green, feel free to review+land as I might not be around by the time the bots finish.

Thanks!
